include/exadg/operators/navier_stokes_calculators.cpp

```diff
@@ -84,10 +144,232 @@ ShearRateCalculator<dim, Number>::cell_loop(dealii::MatrixFree<dim, Number> cons
+{
+  dst = 0;
+
+  matrix_free->cell_loop(&This::cell_loop, this, dst, src);
+}
+
+template<int dim, typename Number>
+void
+MagnitudeCalculator<dim, Number>::cell_loop(dealii::MatrixFree<dim, Number> const & matrix_free,
+                                            VectorType &                            dst,
+                                            VectorType const &                      src,
+                                            Range const & cell_range) const
+{
+  IntegratorVector integrator_vector(matrix_free, dof_index_u, quad_index);
+  IntegratorScalar integrator_scalar(matrix_free, dof_index_u_scalar, quad_index);
+
+  for(unsigned int cell = cell_range.first; cell < cell_range.second; ++cell)
+  {
+    integrator_vector.reinit(cell);
+    integrator_vector.gather_evaluate(src, dealii::EvaluationFlags::values);
+
+    integrator_scalar.reinit(cell);
+
+    for(unsigned int q = 0; q < integrator_scalar.n_q_points; q++)
+    {
+      scalar magnitude = integrator_vector.get_value(q).norm();
+      integrator_scalar.submit_value(magnitude, q);
+    }
+    integrator_scalar.integrate_scatter(dealii::EvaluationFlags::values, dst);
+  }
+}
+
+template<int dim, typename Number>
+QCriterionCalculator<dim, Number>::QCriterionCalculator()
+  : matrix_free(nullptr),
+    dof_index_u(0),
+    dof_index_u_scalar(0),
+    quad_index(0),
+    compressible_flow(false)
+{
+}
+
+template<int dim, typename Number>
+void
+QCriterionCalculator<dim, Number>::initialize(
+  dealii::MatrixFree<dim, Number> const & matrix_free_in,
+  unsigned int const                      dof_index_u_in,
+  unsigned int const                      dof_index_u_scalar_in,
+  unsigned int const                      quad_index_in,
+  bool const                              compressible_flow_in)
+{
+  matrix_free        = &matrix_free_in;
+  dof_index_u        = dof_index_u_in;
+  dof_index_u_scalar = dof_index_u_scalar_in;
+  quad_index         = quad_index_in;
+  compressible_flow  = compressible_flow_in;
+}
+
+template<int dim, typename Number>
+void
+QCriterionCalculator<dim, Number>::compute(VectorType & dst, VectorType const & src) const
+{
+  dst = 0;
+
+  matrix_free->cell_loop(&This::cell_loop, this, dst, src, compressible_flow);
+}
+
+template<int dim, typename Number>
+void
+QCriterionCalculator<dim, Number>::cell_loop(dealii::MatrixFree<dim, Number> const & matrix_free,
+                                             VectorType &                            dst,
+                                             VectorType const &                      src,
+                                             Range const & cell_range) const
+{
+  CellIntegratorVector integrator_vector(matrix_free, dof_index_u, quad_index);
+  CellIntegratorScalar integrator_scalar(matrix_free, dof_index_u_scalar, quad_index);
+
+  for(unsigned int cell = cell_range.first; cell < cell_range.second; ++cell)
```

Comment:
if statement before the loop or inside? minor code duplication or not?

Reply:
i agree with the present version